Update tutorial to use dynamic state for viewport and scissor

SaschaWillems opened the pull request:
This PR updates the tutorial to use dynamic state for viewport and scissor rectangles. This simplifies things a lot and no longer requires e.g. pipelines and render passes to be recreated with the swap chain on window resize.

* Update chapter "Fixed function":
  * Better wording on baked state in pipeline state objects
  * Moved dynamic state paragraph to the top and also slightly reworded it
  * Updated the viewport and scissor state chapter with dynamic state and why it's better to use this instead of baking it
* Update chapter "Command buffers"
  * Added viewport and scissor setup to command buffer recording
* Update chapter "Swapchain recreation"
  * With dynamic state for viewport and scissor, it's no longer required to recreate pipeline and render pass when the swapchain is recreated
* Update all code files that had viewport and scissor state baked into the pipeline to use dynamic state instead and set viewport and scissor at command buffer creation
* Update all code files after the swap chain chapter to no longer recreate the render pass and pipeline when the swap chain was recreated

Project card: https://github.com/KhronosGroup/VulkanTutorial/projects/1#card-69148058

SaschaWillems committed:
Prefer use of dynamic state for viewport and scissor

SaschaWillems committed:
Slight rewording, working in-page anchor

SaschaWillems committed:
Updated command buffers chapter and code to use dynamic viewport and scissor state

SaschaWillems committed:
Updated swap chain recreation chapter

No more need to recreate the graphics pipeline as viewport and scissor state are now dynamic

SaschaWillems committed:
Updated command buffer chapter code to use dynamic viewport and scissor

SaschaWillems committed:
Removed render pass destruction/creation for swap chain recreation

This is not needed anymore

SaschaWillems committed:
Use dynamic state for viewport and scissor for samples after the fixed function chapter

SaschaWillems committed:
Use dynamic state for viewport and scissor for samples after the swap chain recreation chapter

Removed render pass destruction/creation for swap chain recreation, no longer necessary

SaschaWillems committed:
Removed duplicate code

SaschaWillems committed:
Wording change

SaschaWillems committed:
Changed wording on chapter intro, moved dynamic state paragraph to the top

charles-lunarg commented on code/10_fixed_functions.cpp:
```diff
@@ -379,24 +379,10 @@ class HelloTriangleApplication {
         inputAssembly.topology = VK_PRIMITIVE_TOPOLOGY_TRIANGLE_LIST;
         inputAssembly.primitiveRestartEnable = VK_FALSE;
 
-        VkViewport viewport{};
-        viewport.x = 0.0f;
-        viewport.y = 0.0f;
-        viewport.width = (float) swapChainExtent.width;
-        viewport.height = (float) swapChainExtent.height;
-        viewport.minDepth = 0.0f;
-        viewport.maxDepth = 1.0f;
```
Even though dynamic viewport should make this unecessary, it was my understanding that it was still needed. That or I am remembering a driver or validation layer bug and this was the fix for it.

SaschaWillems replied:
I'm pretty sure that is not needed when using dynamic states. Telling the pipeline that we use one viewport and one scissor is sufficient.

gpx1000 commented on en/03_Drawing_a_triangle/02_Graphics_pipeline_basics/02_Fixed_functions.md:
```diff
@@ -1,7 +1,35 @@
-the structures to configure these fixed-function operations.
+graphics pipeline. In Vulkan you have to be explicit about most pipeline state as
+it'll be baked into an immutable pipeline state object. In this chapter we'll fill 
+in all of the structures to configure these fixed-function operations.
```
... about most pipeline states ...  <-- add the plural

gpx1000 commented on en/03_Drawing_a_triangle/02_Graphics_pipeline_basics/02_Fixed_functions.md:
```diff
@@ -1,7 +1,35 @@
+
+While *most* of the pipeline state needs to be baked into the pipeline state, 
+a limited amount of the state *can* actually be changed without recreating the 
+pipeline at draw time instead. Examples are the size of the viewport, line width 
```
remove instead from pipeline at draw time.

gpx1000 reviewed (approved):
minor grammar edits.

charles-lunarg commented on en/03_Drawing_a_triangle/02_Graphics_pipeline_basics/02_Fixed_functions.md:
````diff
@@ -100,20 +128,55 @@ viewport.
+viewportState.scissorCount = 1;
+```
+
+Viewport(s) and scissor rectangle(s) are then set in the command buffer before issuing the draw commands:
````
Since command buffers haven't been introduced yet as a concept, I think this needs some additional work. Probably wise to not try to introduce command buffers here and instead link to their chapter saying "we will cover this later".

```
Since we are not setting the values of the viewport(s) and scissor rectangle(s) in pipeline creation, we will have to set the data when we issue draw calls. While we cover that in detail in chapter (INSERT CHAPTER LINK), the code outline looks like the following

vkCmdSetViewport(commandBuffer, 0, 1, &viewport);
vkCmdSetScissor(commandBuffer, 0, 1, &scissor);
```

SaschaWillems replied:
Good point, but command buffers are already introduced (or at least mentioned)) in chapter 01, so I think it's fine to mention them here. Having a link to the command buffer chapter is a good idea though.

charles-lunarg commented on en/03_Drawing_a_triangle/02_Graphics_pipeline_basics/02_Fixed_functions.md:
````diff
@@ -100,20 +128,55 @@ viewport.
+vkEndCommandBuffer(commandBuffer);
+```
+
+With dynamic state it's even possible to specify different viewports and or scissor rectangles within a single command buffer.
````
Again, without knowing what a command buffer is, this sentence makes no sense to beginners.

charles-lunarg commented on en/03_Drawing_a_triangle/02_Graphics_pipeline_basics/02_Fixed_functions.md:
```diff
@@ -100,20 +128,55 @@ viewport.
+
+With dynamic state it's even possible to specify different viewports and or scissor rectangles within a single command buffer.
+
+Without dynamic state, viewport and scissor rectangle need to be combined into a viewport
+state for the pipeline using the `VkPipelineViewportStateCreateInfo` struct. This makes the viewport and scissor rectangle setup for this pipeline immutable, so any changes required to these would also require the pipeline to be recreated.
```
```suggestion
Without dynamic state, the viewport and scissor rectangle need to be set in the pipeline using the `VkPipelineViewportStateCreateInfo` struct. This makes the viewport and scissor rectangle for this pipeline immutable.
Any changes required to these values would require a new pipeline to created with the new values.
```

Cleaned up some of the language I felt.
Specifically I replaced `recreated` with 'create a new pipeline' since some users might think there is a specific Vulkan function for that. Feel free to disagree and keep the original language. (I might be nitpicking just a little too hard here).

charles-lunarg commented on en/03_Drawing_a_triangle/03_Drawing/01_Command_buffers.md:
```diff
@@ -276,8 +276,29 @@ vkCmdBindPipeline(commandBuffer, VK_PIPELINE_BIND_POINT_GRAPHICS, graphicsPipeli
+VkViewport viewport{};
+viewport.x = 0.0f;
+viewport.y = 0.0f;
+viewport.width = (float) swapChainExtent.width;
+viewport.height = (float) swapChainExtent.height;
```
prefer using `static_cast<float>(swapChainExtent)`?  (general C++ best practices)

SaschaWillems replied:
Good point. The rest of the tutorial also uses the C++ style casts. I have changed that.

charles-lunarg commented on en/03_Drawing_a_triangle/04_Swap_chain_recreation.md:
````diff
@@ -18,23 +18,15 @@ void recreateSwapChain() {
 }
 ```
 
 We first call `vkDeviceWaitIdle`, because just like in the last chapter, we
 shouldn't touch resources that may still be in use. Obviously, the first thing
 we'll have to do is recreate the swap chain itself. The image views need to be
````
```suggestion
We first call `vkDeviceWaitIdle`, because just like in the last chapter, we
shouldn't touch resources that may still be in use. Obviously, we'll have
recreate the swap chain itself. The image views need to be
```
(just noticed the language here says 'first' twice, which is redundant and a bit odd)

charles-lunarg commented on en/03_Drawing_a_triangle/04_Swap_chain_recreation.md:
```diff
@@ -18,23 +18,15 @@ void recreateSwapChain() {
-also needs to be rebuilt. It is possible to avoid this by using dynamic state
-for the viewports and scissor rectangles. Finally, the framebuffers directly
-depend on the swap chain images.
+recreated because they are based directly on the swap chain images. Finally, the 
+framebuffers directly depend on the swap chain images.
```
```suggestion
recreated because they are based directly on the swap chain images. Finally, the 
framebuffers directly depend on the swap chain images, and thus must be recreated as well.
```

charles-lunarg commented on en/03_Drawing_a_triangle/04_Swap_chain_recreation.md:
```diff
@@ -53,8 +45,6 @@ void recreateSwapChain() {
 
     createSwapChain();
     createImageViews();
-    createRenderPass();
```
I'm not actually sure we want to get rid of renderpass re-creation. 
The reason is that there is a non-zero chance of the Swapchain Image Format changing. I can imagine a weird setup where the window gets moved between a HDR and SDR monitor, and the window system requests a format change. Or something equally ridiculous. On windows and x11, I think these are unlikely but with wayland and more exotic window managers I don't have such confidence. 

I may be raising an issue that simply doesn't exist out in the wild, but its one thing that I would love some reassurance that "it doesn't happen in practice" so that this code isn't very subtly wrong.

SaschaWillems replied:
That's a good point, but tbh in all my years doing Vulkan I have never encountered such a situation, so I'd prefer to not have to recreate the render pass here. I don't do this in my samples either.

Adding notes on artificial cases may also irritate readers.

To sum it up: I'm not sure on this and open for discussion :)

SaschaWillems replied:
As per today's call: I'll reimplement this to make sure the sample is 100% correct, even in special cases.

charles-lunarg replied:
Apologies for making more work for you, but I think removing the recreation of renderpasses is okay as long as it is noted that we use swapchain image formats which are pretty much always supported. If a different format is used (eg, HDR) care is needed to ensure that the formats being used are always correct, no trying to render SDR to an HDR swapchain and vice versa.

SaschaWillems replied:
No problem at all :) I'll add a note on when render pass re-creation may be necessary, so for those few occasions where it can happen people at least have a pointer as to why it could happen.

SaschaWillems replied:
I have added a note on renderpass recreation:

> Note that we don't recreate the renderpass for simplicity. In theory it can be possible for the swap chain image format to change during an applications' life-time, e.g. when moving a window from an standard range to an high dynamic range monitor. This may require the application to recreate the renderpass to make sure the change between dynamic ranges is properly reflected.

Hope that fits into the tutorial and is easy (enough) to understand. Let me know if you're fine with that :)

charles-lunarg reviewed (commented):
Only really concerned with the use of command buffers before they are introduced (in a later chapter) as far as an issue I have with the changes.

Overv committed:
Update PDFs (fixes #279)

SaschaWillems committed:
Fixed wording

SaschaWillems committed:
Fixed wording

SaschaWillems committed:
C++ style cast

SaschaWillems committed:
Wording

SaschaWillems committed:
Wording

SaschaWillems committed:
Wording

Commit:
Redundant word removed (#282)

Overv committed:
Fix typo (fixes #283)

rHermes committed:
Remove outdated reference to swapchain (#285)

This commit removes leftover references to the number of swapchain
images in the "Uniform buffers" chapter.

SaschaWillems committed:
Added note on renderpass recreation

charles-lunarg reviewed (approved):
Thanks! I'm sure if I review it again i'll nitpick language, but this has addressed all the major concerns I have.

Overv committed:
Fix syntax error & inconsistency in sample code

rHermes committed:
Fix small problem in "Depth buffering" (#289)

This code was still refering to `createCommandBuffers`  instead of `recordCommandBuffer`.

Overv committed:
Fix parts of the tutorial that weren't update to match the code (fixes #288)

Overv committed:
Fix inconsistency between tutorial and code

SaschaWillems committed:
Merge branch 'master' into dynamic_viewport

SaschaWillems committed:
Apply suggestions from code review

Incorporate suggestions

Co-authored-by: Alexander Overvoorde <overv161@gmail.com>

SaschaWillems committed:
Added dynamic state to conclusion chapter

gpx1000 reviewed (approved):
LGTM

SaschaWillems committed:
Remove unnecessary link

SaschaWillems committed:
Remove code referencing command buffers too early

SaschaWillems commented:
Alexander has merged this PR into the Vulkan Tutorial repo, so I'm closing this one.
